packages/legend-query/src/stores/QueryFunctionsState.ts

```diff
@@ -0,0 +1,468 @@
+  const rootIds: string[] = [];
+  const nodes = new Map<string, QueryBuilderPackageElementTreeNodeData>();
+  const validDisplayablePackageSet = getValidDisplayablePackageSet(
+    queryBuilderState,
+    rootPackageName,
+  );
+  switch (rootPackageName) {
+    case ROOT_PACKAGE_NAME.PROJECT_DEPENDENCY_ROOT:
+      if (
+        queryBuilderState.graphManagerState.graph.dependencyManager.functions
+          .length === 0
+      ) {
+        return { rootIds, nodes };
+      }
+      break;
+    default:
+      if (queryBuilderState.graphManagerState.graph.ownFunctions.length === 0) {
+        return { rootIds, nodes };
+      }
+  }
+
+  root.children
+    .slice()
+    .filter((child) => !(child instanceof Unit))
+    .filter(
+      (child) =>
+        child instanceof Package && validDisplayablePackageSet.has(child),
+    )
+    .sort((a, b) => a.name.localeCompare(b.name))
+    .sort(
+      (a, b) => (b instanceof Package ? 1 : 0) - (a instanceof Package ? 1 : 0),
+    )
+    .forEach((childPackage) => {
+      const childTreeNodeData = generateQueryBuilderPackableElementTreeNodeData(
+        queryBuilderState,
+        childPackage,
+        rootPackageName,
+      );
+      addUniqueEntry(rootIds, childTreeNodeData.id);
+      nodes.set(childTreeNodeData.id, childTreeNodeData);
+    });
+  return { rootIds, nodes };
+};
+
+export const getTreeChildNodes = (
+  queryBuilderState: QueryBuilderState,
+  node: QueryBuilderPackageElementTreeNodeData,
+  data: TreeData<QueryBuilderPackageElementTreeNodeData>,
+  rootPackageName = ROOT_PACKAGE_NAME.MAIN,
+): QueryBuilderPackageElementTreeNodeData[] => {
+  if (node.childrenIds && node.packageableElement instanceof Package) {
+    populateQueryBuilderPackageElementTreeNodeChildren(
+      queryBuilderState,
+      node,
+      data,
+      rootPackageName,
+    );
+    return node.childrenIds
+      .map((id) => data.nodes.get(id))
+      .filter(isNonNullable)
+      .sort((a, b) => a.label.localeCompare(b.label))
+      .sort(
+        (a, b) =>
+          (b.packageableElement instanceof Package ? 1 : 0) -
+          (a.packageableElement instanceof Package ? 1 : 0),
+      );
+  }
+  return [];
+};
+
+const getAllPackagesFromElement = (element: PackageableElement): Package[] => {
+  if (element.package) {
+    return [element.package].concat(getAllPackagesFromElement(element.package));
+  }
+  return [];
+};
+
+export const generateFunctionSignature = (
+  element: ConcreteFunctionDefinition,
+  fullPath: boolean,
+): string =>
+  `${fullPath ? element.path : element.name}(${element.parameters
+    .map(
+      (p) =>
+        `${p.name}: ${p.type.value.name}${getSimpleMultiplicityDescription(
+          p.multiplicity,
+        )}`,
+    )
+    .join(', ')})` +
+  `: ${element.returnType.value.name}${getSimpleMultiplicityDescription(
+    element.returnMultiplicity,
+  )}`;
+
+export const generateDefaultValueForPrimitiveParameter = (
```

Comment:
Can we reuse this from somewhere?

Reply:
Tried to use `generateDefaultValueForPrimitiveType`, but didn't work as i expect